src/main.rs

```diff
@@ -50,29 +51,141 @@ impl MerkleTree {
+    ///     Consider the example:
+    ///                     R
+    ///                 H1      H2      --> adding a new element will require to have 8 leaves (next power of two).
+    ///             H3  H4      H5 H6
+    /// Therefore another "half tree" is constructed and both are appended converging in a calculated new root
+    ///                                 R'
+    ///                     R1                   R2
+    ///                 H1      H2            H7     H8                 Where everything under R1 is the original tree,
+    ///             H3  H4      H5 H6      H9 H10    H11 H12            everything under R2 is the new subtree and H9 is the hash of the new element
+    ///                                                                 (H10, H11, H12 are default data) and R' is the new calculated root (hash of (R1+R2))
+    ///
+    pub fn add_element(&mut self, new_element: Vec<u8>) {
+        let next_index = self.last_element_index() + 1;
+
+        // Contemplates Case 2. The element requires a new index past the existing ones and the tree must be extended.
+        if next_index == self.leaves_amount() {
+            self.extend_tree(new_element);
+        }
+        // Contemplates Case 1. The new element replaces an existing index and the tree must be updated.
+        else {
+            self.modify_existing_leaf_and_update_parents(new_element);
+        }
+
+        // After any of the cases, the last index with real data must be updated.
+        self.set_last_element_index(next_index);
+    }
+
+    #[allow(clippy::needless_range_loop)] // Allow this to make code more understandable
+    /// Manages the case when the tree must be extended to add an element.
+    fn extend_tree(&mut self, new_element: Vec<u8>) {
+        let leaves_amount = self.leaves_amount();
+        // The needed amount of default data to fill the leaves. Given by the calculation:
+        // Example: if the current tree has 4 leaves then:
+        // current_amount_of_leaves + new_leaf = 4 + 1 = 5
+        // next_power_of_two(5) = 8
+        // needed leaves to reach next power of two from current state -> 8 - 4 = 4
+        // taking into account that 1 of those leaves is the actual new data the fill amount is 3
+        let fill_amount = ((leaves_amount + 1).next_power_of_two() - leaves_amount) - 1;
+
+        // Create the leaves level of the new subtree
+        let mut leaves = vec![sha256::digest(new_element)];
+        leaves.append(&mut vec![DEFAULT_ZERO_HASH.to_string(); fill_amount]);
+
+        // Create a new merkle tree for the new data subset
+        let mut half_merkle_tree = create_merkle_tree_from_leaves(leaves);
+
+        // Append the new merkle tree's levels to the previous tree
+        for i in 0..self.height() {
+            self.merkle_tree[i].append(&mut half_merkle_tree[i]);
```

Comment:
Here you could use the function `get_level(i)`:
```suggestion
            self.merkle_tree.get_level(i).append(&mut half_merkle_tree.get_level(i));
```

Reply:
the thing with get_level() is that it returns an immutable reference

Reply:
Oh I didn't realize that! 
Just if you want to not create the `half_merkle_tree` a `mut` you can use this:
```
// Create a new merkle tree for the new data subset
let half_merkle_tree = create_merkle_tree_from_leaves(leaves);

// Append the new merkle tree's levels to the previous tree
for i in 0..self.height() {
    self.merkle_tree[i].extend_from_slice(&half_merkle_tree[i]);
}
```

Reply:
done! thanks :)